### PR TITLE
Resolve Ambiguous Group-By-Items in Queries

### PR DESCRIPTION
**Pull request by plypaul**
### Description

This PR updates MF to resolve ambiguous group-by-items in queries by incorporating the foundation that was set up in prior PRs:

* Update the `DataflowPlanBuilder` to use a lookup to resolve ambiguous group-by-items in filters.
* Map existing input types (strings that specify group-by-items in different ways, `*QueryParameter` classes) into a set of common intermediate input classes (`ResolverInput*`) that contain spec patters.
* Implement a query resolver that resolves the patterns in input classes into concrete specs.
* Rewrite `MetricFlowQueryParser` to only parse the current inputs into the common intermediate input classes, and use the query resolver to validate inputs and resolve the query items into concrete specs.

In addition, the way that query validation errors were collected and reported has been changed. Insted of raising an exception when an error is encountered, issues are collected and returned together. This makes resolving those errors easier for the user as all errors can be resolved in a single pass vs. fixing one error, re-running the validation to find the next error, etc.

### CODE REVIEW

**Comment by plypaul** [comment]
This PR needs to be rebased against main to resolve the missing snapshot test failure.

File: metricflow/time/time_granularity_solver.py [hunk 1]
```diff
@@ -1,217 +0,0 @@
-from __future__ import annotations
```
Comment (tlento) [inline metricflow/time/time_granularity_solver.py:1]: YAY! Red diff party!

celebrate!

File: metricflow/test/dataflow/builder/test_dataflow_plan_builder.py [hunk 2]
```diff
@@ -430,28 +403,15 @@ def test_where_constrained_with_common_linkable_plan(  # noqa: D
...
-                    PydanticWhereFilter(
-                        where_sql_template="{{ Dimension('listing__country_latest') }} = 'us'",
-                    )
-                )
-            ),
-        )
+    query_spec = query_parser.parse_and_validate_query(
+        metric_names=("bookings",),
+        group_by_names=("listing__country_latest",),
+        where_constraint_str="{{ Dimension('listing__country_latest') }} = 'us'",
```
Comment (tlento) [inline metricflow/test/dataflow/builder/test_dataflow_plan_builder.py:412]: Oh this is so much better...

File: metricflow/test/query/test_query_parser.py [hunk 3]
```diff
@@ -256,13 +252,12 @@ def test_query_parser_case_insensitivity(bookings_query_parser: MetricFlowQueryP
 
 
 def test_query_parser_invalid_group_by(bookings_query_parser: MetricFlowQueryParser) -> None:  # noqa: D
-    with pytest.raises(UnableToSatisfyQueryError):
+    with pytest.raises(InvalidQueryException):
```
Comment (tlento) [inline metricflow/test/query/test_query_parser.py:255]: We probably need to communicate this out so MFS/SLG exception handling for alerts and stuff accounts for this in the same way they account for UnableToSatisfyQueryError.

File: metricflow/test/query_rendering/test_query_rendering.py [hunk 4]
```diff
@@ -95,31 +95,20 @@ def test_filter_with_where_constraint_on_join_dim(
...
-                WhereSpecFactory(
-                    column_association_resolver=column_association_resolver,
-                ).create_from_where_filter(
-                    PydanticWhereFilter(
-                        where_sql_template="{{ Dimension('listing__country_latest') }} = 'us'",
-                    )
-                )
-            ),
-        )
+    query_spec = query_parser.parse_and_validate_query(
```
Comment (tlento) [inline metricflow/test/query_rendering/test_query_rendering.py:104]: YES!

File: metricflow/test/query_rendering/test_query_rendering.py [hunk 5]
```diff
@@ -95,31 +95,20 @@ def test_filter_with_where_constraint_on_join_dim(
...
     consistent_id_object_repository: ConsistentIdObjectRepository,
     sql_client: SqlClient,
 ) -> None:
     """Tests converting a dataflow plan to a SQL query plan where there is a join between 1 measure and 2 dimensions."""
-    dataflow_plan = dataflow_plan_builder.build_plan(
-        MetricFlowQuerySpec(
-            metric_specs=(MetricSpec(element_name="bookings"),),
-            dimension_specs=(
-                DimensionSpec(
-                    element_name="is_instant",
```
Comment (tlento) [inline metricflow/test/query_rendering/test_query_rendering.py:108]: Oh duh, that's why these changed in the snapshots.

**Review by tlento — APPROVED** [review]
spongebob done

That took some time. Anyway, I left a ton of commentary in line. As with the others, it's all either minor stuff or things we should probably be addressing separately. This PR, in particular, is gnarly enough as it is, and since the test outputs remained consistent (with the exception of expected mechanical changes to some dataflow plan files) I think it's reasonably safe to say our behaviors haven't changed.

Let's get these merged so we can start doing user-facing testing!

File: metricflow/dataflow/builder/dataflow_plan_builder.py [hunk 6]
```diff
@@ -686,10 +749,11 @@ def build_computed_metrics_node(
 
     def _build_input_measure_spec_for_base_metric(
         self,
+        filter_spec_factory: WhereSpecFactory,
         metric_reference: MetricReference,
-        column_association_resolver: ColumnAssociationResolver,
         child_metric_offset_window: Optional[MetricTimeWindow],
         child_metric_offset_to_grain: Optional[TimeGranularity],
+        descendent_filter_specs: Sequence[WhereFilterSpec],
```
[the remark this answers is not in the document]
Comment (plypaul) [reply metricflow/dataflow/builder/dataflow_plan_builder.py:772]: lol seems like there's not a clear answer for that spelling. That's the best I got so far though.

File: metricflow/dataflow/builder/dataflow_plan_builder.py [hunk 7]
```diff
@@ -686,10 +749,11 @@ def build_computed_metrics_node(
...
         query_contains_metric_time: bool,
         culmination_description: Optional[CumulativeMeasureDescription],
     ) -> MetricInputMeasureSpec:
```
[the remark this answers is not in the document]
Comment (plypaul) [reply metricflow/dataflow/builder/dataflow_plan_builder.py:775]: Updated.

**Pull request merged**